DDP-5469: Add browser JS error monitoring to all apps (patient-facing and DSM)

fat23cat opened the pull request:
This PR covers only patient-facing apps, if this stuff looks good, we can implement this for DSM.
Please note that our custom error handler will track **only** JS errors, for instance: if we try to run some value as a function, but it's not a function; if we try to get an object property, but the object is undefined, etc. It won't track network errors, server errors, or bad requests.
In case we want explicitly pass some error to the GCP console, we can inject StackdriverErrorReporterService into service or component, and happened something which we want to track, we can generate a new Error and pass to handleError method.
You can observe errors here: https://console.cloud.google.com/errors?time=PT1H&order=COUNT_DESC&resolution=OPEN&resolution=ACKNOWLEDGED&project=broad-ddp-dev&authuser=1&folder&organizationId&supportedpurview=project

fat23cat committed:
Install stackdriver

fat23cat committed:
Implement service

fat23cat committed:
Update config script

fat23cat committed:
Use new service

fat23cat committed:
Remove key

fat23cat commented on config/pepperConfig.js.ctmpl:
```diff
@@ -9,6 +9,8 @@ var DDP_ENV = {
     doLocalRegistration: false,
     {{ with secret (printf "secret/pepper/%s/%s/conf" $environment $version) }}
     mapsApiKey: "{{.Data.mapsApiKey}}",
+    errorReportingApiKey: "{{.Data.errorReportingApiKey}}",
+    projectGcpId: "{{.Data.projectGcpId}}",
```
I need some assistance here: I think we should store the key and the projectId in the vault, but I don't know how to put data there. In addition to that, we need to generate different keys for dev, test, staging, and prod, but I have access only to the dev GCP project.
We need to generate keys for test, staging, and prod, using the following example: https://console.cloud.google.com/apis/credentials/key/bc5b7470-db88-4132-a609-3b16164938df?authuser=1&project=broad-ddp-dev&supportedpurview=project

MocanaAtBroad replied:
We have this value in Vault already:
```suggestion
    projectGcpId: "{{.Data.googleProjectId}}",
```

fat23cat commented on ddp-workspace/projects/ddp-testboston/src/app/app.module.ts:
```diff
@@ -97,6 +98,8 @@ sdkConfig.doLocalRegistration = DDP_ENV.doLocalRegistration;
 sdkConfig.mapsApiKey = DDP_ENV.mapsApiKey;
 sdkConfig.auth0Audience = DDP_ENV.auth0Audience;
 sdkConfig.projectGAToken = DDP_ENV.projectGAToken;
+sdkConfig.errorReportingApiKey = DDP_ENV.errorReportingApiKey;
```
I've accidentally pushed errorReportingApiKey to the github repo, the key was re-generated.

zyme replied:
Not a big deal--this will show up in the frontend deployment anyway, and we will need to lock it down by origin in the GCP console anyway.  We'll need to generate 4 env specific keys for this...

MocanaAtBroad commented:
The setup that includes the reporter script and hooking it up to Angular looks good!
I think we need a couple of changes to make it useful for us:
1) Make the logging of errors to GCP conditional. Probably easiest way to do this is to just have a boolean flag in pepperConfig.js (e.g., doGcpErrorReporting=true). Allows us turn it off when working locally.
2) Have reports include the service name (i.e., specify that it comes ddp-testboston). Right now errors are reported as just coming from "web" which makes error less useful.
3) Wire the error reporting into our error logging. So if we call LoggingService.logError , whatever would have been sent to the console gets included in the error report (file, line, stack trace).
4) Include userGuid (if available)
5) Browser info

Based on payload specified in docs, all this ought to be possible:
https://cloud.google.com/error-reporting/docs/formatting-error-messages#json_representation

MocanaAtBroad reviewed (commented):
I have added `errorReportingApiKey` for dev to Vault. As noted in my comments. If you apply my suggested change to pepperConfig.js.ctmpl file , we should be generating a valid config file the needed values.
Note that I also tweaked API key configuration in GCP to limit referers to be localhost:4200/* and *.dev.datadonationplatform.org:/*.

fat23cat committed:
Update config/pepperConfig.js.ctmpl

Co-authored-by: MocanaAtBroad <mocana@broadinstitute.org>

alextanlan committed:
Merge branch 'develop' into DDP-5469_error-handling

zyme commented on ddp-workspace/package-lock.json:
```diff
@@ -5209,6 +5209,14 @@
+      "resolved": "https://registry.npmjs.org/error-stack-parser/-/error-stack-parser-2.0.6.tgz",
+      "integrity": "sha512-d51brTeqC+BHlwF0BhPtcYgF5nlzf9ZZ0ZIUQNZpc9ZB9qw5IJ2diTrBY9jlCJkTLITYPjmiX6OWCwH+fuyNgQ==",
+      "requires": {
+        "stackframe": "^1.1.1"
```
Can you pin the versions to an explicit version rather than `^` throughout?  We try to fix versions whenever possible so that we have reproducible builds.

alextanlan replied:
Seems to me that it is not worth doing because they are peer-dependencies.
At the same time, all dependencies in our `package.json` are explicitly set.   
What are your thoughts about that ?

zyme commented:
I'm curious about the current limitations for this.  Would we pick up on our internal loading of the `error` page and unexpected API codes like `500`?

alextanlan committed:
Provide StackdriverErrorReporterService in DDP module only

  as other app modules import ddpModule

alextanlan committed:
Make the logging of JS errors to GCP conditional

alextanlan committed:
Add service name into reports

alextanlan committed:
Add user info into reports

alextanlan committed:
Remove browser info

  because the user agent info exists in GCP Error report by default

alextanlan committed:
Wire the GCP error reporting into our error logging

  e.g. if we call LoggingService.logError,
  the error is sent to GCP error report as well

alextanlan committed:
Fix tests

alextanlan commented:
> I think we need a couple of changes to make it useful for us:
> 
> 1. Make the logging of errors to GCP conditional. Probably easiest way to do this is to just have a boolean flag in pepperConfig.js (e.g., doGcpErrorReporting=true). Allows us turn it off when working locally.
> 2. Have reports include the service name (i.e., specify that it comes ddp-testboston). Right now errors are reported as just coming from "web" which makes error less useful.
> 3. Wire the error reporting into our error logging. So if we call LoggingService.logError , whatever would have been sent to the console gets included in the error report (file, line, stack trace).
> 4. Include userGuid (if available)
> 5. Browser info

- All points above are done

alextanlan commented:
> I'm curious about the current limitations for this. Would we pick up on our internal loading of the `error` page and unexpected API codes like `500`?

We can send any custom error/message to GCP Error report from any custom error handler/any code place,
e.g. just before redirecting to the error page.

zyme commented:
This is looking good!  Can we deploy it to a dev study and deliberately break stuff so we can confirm that messages end up in SD?

alextanlan committed:
Fix message

alextanlan committed:
Test purpose commit

alextanlan committed:
Remove debug code

alextanlan commented:
> Can we deploy it to a dev study and deliberately break stuff so we can confirm that messages end up in SD?

Tested on https://brain.dev.datadonationplatform.org. 
Errors are sent to SD.

alextanlan committed:
Merge branch 'develop' into DDP-5469_error-handling

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/services/configuration.service.ts:
```diff
@@ -58,4 +61,7 @@ export class ConfigurationService {
     rtlLanguages: string[] = [];
+    // The global flag indicates whether we need report JS errors to Google Cloud Error Report API
+    // It can be overridden for any single app (in proper app.module.ts)
+    doGcpErrorReporting = true;
```
Can we make sure that we can set this configuration value by modifying the generated pepperConfig.js file?
Easiest way I can think of is by adding boolean configuration parameter to pepperConfig.js.ctmpl and reading in app.module.ts. Could also set `doGcpErrorReporting` when one of the two logging config parameters are missing. Up to you.
When running app locally most of the time we will not want to send errors to GCP. I could hand-edit this file, but git will identify it as a modified file. Rather edit pepperConfig.js that is not kept in the repo.
Take look at app.module.ts file for how we have been reading pepperConfig.js to set configuration values.

alextanlan replied:
Got it.
According to the point, it's fixed in 45027f3245726a8a1cc8213990df2a161f471922, f41d777ac0b00b78b7ad1b3f10048962fcb2d60f

MocanaAtBroad reviewed (commented):
It is looking good! One small requested change.

alextanlan committed:
Move doGcpErrorReporting flag to pepperConfig

alextanlan committed:
Init doGcpErrorReporting flag in app modules

  get it from pepperConfig
  for angio, brain, mbc, mpc, osteo, testboston

alextanlan committed:
Merge branch 'develop' into DDP-5469_error-handling

alextanlan committed:
Init doGcpErrorReporting for ddp-esc project

MocanaAtBroad commented on config/pepperConfig.js.ctmpl:
```diff
@@ -9,6 +9,17 @@ var DDP_ENV = {
+        doGcpErrorReporting: false,
+    {{else}}
+        doGcpErrorReporting: true,
+    {{end}}
```
Was looking for something simpler. Just want to be able to turn it off locally by editing the local file.
```suggestion
    // The flag indicates whether we need report JS errors to Google Cloud Error Report API
        doGcpErrorReporting: true,
```

alextanlan replied:
Fixed in d1950e1823c3f8ad2aaf348bcfdaa441d1ac5403

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/services/stackdriverErrorReporter.service.ts:
```diff
@@ -0,0 +1,40 @@
+import { Injectable, Inject, ErrorHandler } from '@angular/core';
+import { ConfigurationService } from './configuration.service';
+import {SessionMementoService} from './sessionMemento.service';
```
Getting whitespace tslint warning for this line. Please make sure you turn on tslint in your IDE.
(We probably should add it to the builds too?)
```suggestion
import { SessionMementoService } from './sessionMemento.service';
```

alextanlan replied:
My bad, thank you.

Moreover, I'm going to create a separate PR with global linter errors/warnings fixes.

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/services/logging.service.ts:
```diff
@@ -1,18 +1,27 @@
 import { Injectable, Inject } from '@angular/core';
 import { ConfigurationService } from './configuration.service';
 import { LogLevel } from '../models/logLevel';
+import {StackdriverErrorReporterService} from './stackdriverErrorReporter.service';
```
whitespace tslint warning:
```suggestion
import { StackdriverErrorReporterService } from './stackdriverErrorReporter.service';
```

alextanlan replied:
Fixed.

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/services/logging.service.spec.ts:
```diff
@@ -2,30 +2,33 @@ import { TestBed } from '@angular/core/testing';
 import { ConfigurationService } from './configuration.service';
 import { LoggingService } from './logging.service';
 import { LogLevel } from '../models/logLevel';
+import {StackdriverErrorReporterService} from './stackdriverErrorReporter.service';
```
Whitespace tslint warning.

alextanlan replied:
Fixed.

MocanaAtBroad reviewed (commented):
Included one small code change suggestion and a few tslint warning corrections.
I am pretty sure there are other whitespace warnings. Please inspect code with tslint.
I believe you can use `ng lint` with flags to run on specific project and files if easier to do from command line.
(I am sure there are other files that will be flagged. Let's not worry about those file for now).

zyme commented on config/pepperConfig.js.ctmpl:
```diff
@@ -9,6 +9,17 @@ var DDP_ENV = {
+    projectGcpId: "{{.Data.googleProjectId}}",
+
+    // The flag indicates whether we need report JS errors to Google Cloud Error Report API
+    {{if not .Data.errorReportingApiKey}}
```
@MocanaAtBroad I'm adding these to test/staging/prod...

zyme commented:
`errorReportingApiKey` values have been added to all four environments.

fat23cat commented on ddp-workspace/projects/ddp-esc/src/app/app.module.ts:
```diff
@@ -77,6 +77,7 @@ sdkConfig.doLocalRegistration = DDP_ENV.doLocalRegistration;
 sdkConfig.mapsApiKey = DDP_ENV.mapsApiKey;
 sdkConfig.auth0Audience = DDP_ENV.auth0Audience;
 sdkConfig.projectGAToken = DDP_ENV.projectGAToken;
+sdkConfig.doGcpErrorReporting = DDP_ENV.doGcpErrorReporting;
```
```suggestion
sdkConfig.errorReportingApiKey = DDP_ENV.errorReportingApiKey;
sdkConfig.projectGcpId = DDP_ENV.projectGcpId;
sdkConfig.doGcpErrorReporting = DDP_ENV.doGcpErrorReporting;
```

alextanlan replied:
added.

fat23cat commented on ddp-workspace/projects/ddp-atcp/src/app/app.module.ts:
```diff
@@ -92,6 +92,8 @@ config.mapsApiKey = DDP_ENV.mapsApiKey;
 config.auth0Audience = DDP_ENV.auth0Audience;
 config.projectGAToken = DDP_ENV.projectGAToken;
 config.defaultLanguageCode = DDP_ENV.defaultLanguageCode ? DDP_ENV.defaultLanguageCode : 'en';
+config.errorReportingApiKey = DDP_ENV.errorReportingApiKey;
+config.projectGcpId = DDP_ENV.projectGcpId;
```
```suggestion
sdkConfig.errorReportingApiKey = DDP_ENV.errorReportingApiKey;
sdkConfig.projectGcpId = DDP_ENV.projectGcpId;
sdkConfig.doGcpErrorReporting = DDP_ENV.doGcpErrorReporting;
```

alextanlan replied:
added.

fat23cat commented on ddp-workspace/projects/ddp-prion/src/app/app.module.ts:
```diff
@@ -201,6 +207,8 @@ config.doLocalRegistration = DDP_ENV.doLocalRegistration;
 config.mapsApiKey = DDP_ENV.mapsApiKey;
 config.auth0Audience = DDP_ENV.auth0Audience;
 config.projectGAToken = DDP_ENV.projectGAToken;
+config.errorReportingApiKey = DDP_ENV.errorReportingApiKey;
+config.projectGcpId = DDP_ENV.projectGcpId;
```
```suggestion
config.errorReportingApiKey = DDP_ENV.errorReportingApiKey;
config.projectGcpId = DDP_ENV.projectGcpId;
config.doGcpErrorReporting = DDP_ENV.doGcpErrorReporting;
```

alextanlan replied:
added.

fat23cat commented on ddp-workspace/projects/ddp-rgp/src/app/app.module.ts:
```diff
@@ -92,6 +92,8 @@ config.doLocalRegistration = DDP_ENV.doLocalRegistration;
 config.mapsApiKey = DDP_ENV.mapsApiKey;
 config.auth0Audience = DDP_ENV.auth0Audience;
 config.projectGAToken = DDP_ENV.projectGAToken;
+config.errorReportingApiKey = DDP_ENV.errorReportingApiKey;
+config.projectGcpId = DDP_ENV.projectGcpId;
```
```suggestion
config.errorReportingApiKey = DDP_ENV.errorReportingApiKey;
config.projectGcpId = DDP_ENV.projectGcpId;
config.doGcpErrorReporting = DDP_ENV.doGcpErrorReporting;
```

alextanlan replied:
added.

alextanlan committed:
Add missed config params for doGcpErrorReporting

  for rgb, esc, atcp, prion studies

alextanlan committed:
Fix tslint errors for ddp-esc

alextanlan committed:
Fix some tslint errors

alextanlan committed:
Move check of missing StackDriver params into StackDriver service

  and fix pepperConfig as to have a simple flag
  for sending errors to Google Cloud Error Report

MocanaAtBroad reviewed (approved):
Looks good!
